ENH: Adding Transit Periodogram to astropy.stats

dfm opened the pull request:
Hi all,

This pull request adds a new feature that I've been working on for a while now with contributions from @mirca and @joeldhartman. This feature is another popular astronomy-specific periodogram like Lomb-Scargle, but this one is used to find transiting exoplanets and eclipsing binary systems by using a top-hat basis instead of sines. For historical reasons, this algorithm is often referred to as "box least squares", but I think that the more descriptive "transit periodogram" name is better.

This algorithm was proposed by [Kovács et al. (2002)](https://arxiv.org/abs/astro-ph/0206099) and the method has since become the standard method of detecting transiting planets. In many cases, this is achieved by wrapping the Fortran code released by those authors (many people use [f2py bindings that I wrote](https://github.com/dfm/python-bls) almost 5 years ago). With K2 data continuing to roll in and the launch of TESS (hopefully today! 🤞) it seems like it would be timely to have an implementation of this algorithm within AstroPy.

With this in mind, my collaborators and I have written an efficient implementation of this algorithm in C with Cython bindings that expose an interface that will be familiar for anyone who uses the LombScargle class in astropy.stats.

A build of the documentation for this pull request [can be found here](http://dfm.io/astropy/stats/transit_periodogram.html). I have also written up a tutorial that uses this code to detect planets using data from NASA's K2 Mission that [can be found here](https://gist.github.com/dfm/3a500f3062aa4af72fe8b4a6d604e58d).  I expect that something like this would eventually be submitted to astropy-tutorials, but I thought that it might be useful to have it here for now.

I have discussed this pull request with several AstroPy maintainers (e.g. @eteq and @adrn) and they endorsed the idea of submitting this to core, but if the community would prefer to keep this as a separate package, I'm certainly open to that idea. Either way, I would definitely appreciate any feedback.

Thanks!

pllim commented:
Probably need a rebase to get rid of `astropy-helpers` diff. Need a change log. Might need some squashing too.

pllim commented:
p.s. Usually tests are in their own `tests` sub-directly, separated from code.

dfm commented:
@pllim: thanks for the suggestions! I hear you on the squashing :-) I figured that I would squash after getting feedback, but I'm happy to do it now if you'd prefer. I've moved the tests, added a changelog entry, and (hopefully) fixed the compilation bug with old C compilers.

I'm not actually sure how to fix the astropy helpers issue. I tried rebasing and it says that everything is up to date. Any tips? Thanks!

pllim commented:
@dfm did you do a `git submodule update`?

pllim commented:
Also saw this error in 32-bit build:
```
astropy/stats/transit_periodogram/transit_periodogram.c: In function ‘run_transit_periodogram’:
astropy/stats/transit_periodogram/transit_periodogram.c:169:30: error: ‘INFINITY’ undeclared (first use in this function)
         best_objective[p] = -INFINITY;
                              ^
astropy/stats/transit_periodogram/transit_periodogram.c:169:30: note: each undeclared identifier is reported only once for each function it appears in
error: command 'gcc' failed with exit status 1
```

dfm commented:
I hadn't done that, but I did previously rebase. When I do a `git submodule update` now, it doesn't seem to do anything... submodules have always confused me :-\ sorry!

pllim commented:
I suspect whatever you rebased against was also outdated. Try this...

If you don't have a remote pointing to `astropy/astropy` already, do this (if SSH) or otherwise change the remote URL to use HTTPS:
```
git remote add upstream git@github.com:astropy/astropy.git 
```

Then
```
git fetch upstream master
git rebase upstream/master
git submodule update
git status
```

I did something similar with your branch here and I get something like:
```
$ git status
On branch pr7391
Your branch and 'dfm/transit-periodogram' have diverged,
and have 59 and 51 different commits each, respectively.
```

Theoretically I could force push to your PR here, but it would mess up the author listing, so I will refrain.

dfm commented:
I tried running those commands and still no luck - very strange! There's a doctest failure on the 32bit build that I'm also going to need to track down, but I'm done for the day. I'll check back tomorrow!

Thanks again for your help.

bsipocz commented:
@dfm - This is an awesome addition, and I'm sure a large number of people will appreciate to have this readily available in astropy.

However, I would argue that keeping the name BLS is preferable rather than a more generic term. It's a widely known name for it, and makes it unambiguous which algorithm is implemented leaving room for adding other ones later (though probably not to astropy core but a separate package).

@crawfordsm - I'm happy to help out in the review on the python side, though that it may indeed take some time.

dfm commented:
@crawfordsm @bsipocz Thanks! I'm looking forward to working with y'all to get this into good shape.

Re: naming - I've always disliked the name BLS because I think it's a bit misleading, but I'm willing to defer and you're probably right that sticking with the historical name is the right way to go... it's also fewer characters :-)

bsipocz commented:
@dfm @pllim - getting rid of the helpers update should be easy. You do an interactive rebase, and remove the commit "Rebase ab0b750de7adea3348ef6f13402e10492e3f078a" from the list in the editor as that seem to be the only one touching the helpers.

dfm commented:
@bsipocz: thanks! It looks like I fixed the helpers bit.

mhvk commented:
This looks great in itself, but should it really be in `astropy` proper? So far, we've not included anything that was this specific to a sub-field of astronomy. It would seem to make much more sense as an affiliated package, since it is logically combined with also fitting transit lightcurves, including limb darkening, etc.

crawfordsm commented:
@mhvk   With the support expressed by multiple maintainers, that it is based on highly cited work, and that exoplanets is current a major field of astronomy, I would be inclined to include it.  It also seems like  it might be applicable to other fields.

pllim commented:
Just for consideration: If this has _any_ chance to be taken out and put in another affiliated package in the future, we should think carefully. Does benefit outweigh the risk? Will this stay forever? Should this find a home instead in an exoplanet specific library? Lesson learned from `conesearch`, which is now in `astroquery`. :wink:

bsipocz commented:
I feel that while somewhat specialized, this belongs to stats more than in an exoplanet library. While my experiences getting rather obsolete (or just the field is moving too fast), BLS was typically run along with LS and other stats, sharing the same technical issues, etc while it shares less with more detailed, individual transit modeling. 

Currently stats is a bit of a random pile of different algorithms anyway, and we could have already argued for a few of those whether they belong here or not (not to mention cosmology; I know why we keep it in core but one could use the same argument that it's only useful for a subset of astronomers and yet not infrastructure, so it should belong to a separate package).

Maybe the future refactoring would be to factor out parts of stats into a stat package once it grows disproportionally big?

mhvk commented:
@crawfordsm, @bsipocz - there are many major fields in astronomy, and so far astropy has only included things that benefit more than a single field. I'd certainly argue this is the case for cosmology, which really is more like coordinates and time, providing standard tools to map redshift to distance, time, etc. The analogy that could be drawn would be with a cosmology-focussed PR that, e.g., adds statistical analysis tools for  CMB fluctuations or growth of structure, etc. That would serve at least as many people as this PR, but would be, in my opinion, similarly too narrow for inclusion in the core.

p.s. I do agree that `stats`, like `convolution` is an odd module.

StuartLittlefair commented:
Great to see an efficient implementation of this algorithm in Cython!

Just to chime in on the issue of where this sits, the algorithm is widely used outside of exoplanet science. To me that suggests a place in `astropy.stats`, along the lines of the Lomb Scargle Periodogram.

If there was ever a case to move it to an affiliated package, it would probably be to a generalised time-series analysis package such as (stingray)[https://github.com/StingraySoftware/stingray]. But then the LS periodogram was not put there, so presumably this should not be for similar reasons?

bmorris3 commented:
To echo a bunch of earlier comments: I think `stats` is a good place to put it.

eteq commented:
I think this is a fine addition to `astropy.stats`.  Given the comments above that imply it's used in several different (if related) domains of astro, it seems reasonable to me that it should sit in the core.

In general I think of `astropy.stats` as a bit "looser" than some of the other subpackages.  It's sensible to have a bit of a "grab bag" here, because it can be a pain to maintain small astro-specific tools as separate, totally independent tools.

mhvk commented:
OK, I concede!  I still worry about the core astropy package turning into a python version of iraf, aiming to be everything to everyone and becoming a GB monster in the process, but I guess the eventual answer to that may be, as it was for `astropy-helpers, to split up current `astropy` into multiple pieces.

astrofrog commented:
Since this is of use to more than one field (at least exoplanets and binary stars) I'd be ok with including it in the core package so 👍from me provided that the name is indeed made more generic.

However, one idea I would like to put out there is that it might be time to think about having a new dedicated sub-package for time series - this could include for example the core times series object proposed in [APE9](https://github.com/astropy/astropy-APEs/pull/12), and could then include this algorithm as well as moving the Lomb-Scargle there. Just to be clear, I don't think that sub-package should be ``astropy.time`` but instead could be something like ``astropy.timeseries``. 

What do people think about this?

eteq commented:
I see @astrofrog's point here but I don't think we want APE9/time series developments to hold this up. So what about putting this in stats *for now*, but move it to `timeseries` *if* there's more to put there by the time of the next feature freeze?

mhvk commented:
Definitely no reason to hold up the PR for a possible `timeseries` class - as noted, we'd have to move lomb-scargle already. And it doesn't seem all that bad to import from `stats` for stuff in `timeseries` anyway.

bmorris3 commented:
Hey @dfm, the tests should pass after you make a few super-tiny PEP8 tweaks. 

On line 29 of `astropy.stats.BLSResults`:
```
docstring of astropy.stats.BLSResults:29: WARNING: Unexpected indentation.
```
And line 30: 
```
docstring of astropy.stats.BLSResults:30: WARNING: Block quote ends without a blank line; unexpected unindent.
```
And in your docs page: 
```
/home/travis/build/astropy/astropy/docs/stats/index.rst:75: WARNING: toctree contains reference to nonexisting document 'stats/transit_periodogram'
```

eteq commented:
I *think* I did the two remaining necessary doc fixes.

@astrofrog and I (and the coordination committee) discussed this out-of-band we are OK with the idea of putting this in stats now, but with the explicit intent of moving it into a `timeseries` sub-package when and if such a package exists.  Hopefully that'll be before this hits the wild (i.e., 3.1), but if not we'll just alias the stats version to the new location (with a suitable deprecation warning).

pllim commented:
Someone might have to investigate while CircleCI failed to download data.

eteq commented:
I'm pretty sure the circleci failures are completely unrelated so I wouldn't worry about that for this PR.

@crawfordsm @larrybradley or @bmorris3 - were one of you planning to review this further now that we've addressed the where-it-lives question? (Asking because I see no one approved/requested changes)

adrn commented on CHANGES.rst:
```diff
@@ -65,6 +65,9 @@ astropy.samp
 astropy.stats
 ^^^^^^^^^^^^^
 
+- Added ``TransitPeriodogram`` class to compute transit periodograms for the
```
Update class name here

adrn commented:
Sorry to add noise, but I would prefer if the class name was a bit more verbose -- out of context, `BLS` is very few characters, and could be confusing since LS is also an abbreviation for Lomb-Scargle! I would vote to change it to `BoxLeastSquares` since `LombScargle` doesn't have the word "periodogram" in its name either.

pllim commented:
p.s. Any chance some commits can be squashed? "yolo" commit etc seems a little cryptic.

pllim commented:
Still need to get rid of `astropy_helpers` changes. FYI

dfm commented:
@pllim not "still"... "again" :-) I'm bad a computers sometimes. Thanks!

bsipocz commented:
@dfm - I'm sorry, I missed it totally today that there were still helpers issues.

dfm commented:
@bsipocz - I actually managed to mess this up *after* my rebase... I forgot to run `git submodule update` before `git commit -a`. Totally my bad!

bsipocz commented:
@dfm - Ohh, the trick is not to use `-a` ever ;)

gully commented:
I experimented with this tutorial today.  Looks good to me, everything worked well with the tutorial.  I found a teeny-tiny bit of missing documentation:
- `depth_half` appears in the dictionary of results of `compute_stats()` but does not have an affiliated entry in the documentation
- ⬆️ same with `depth_phased`

larrybradley commented:
I'm not going to be able to review this soon.  I'll leave it to @adrn and @bsipocz.

pllim commented:
Looks like it still needs a change log (if there was one before, it is gone now). And is the coverage decrease expected?

bsipocz commented:
@pllim - coveralls is useless with the cython coverage, it's quite unbeliavable all but one file have >90% coverage, but that one is 30%. I'll try to have a look at it (but can't give an ETA, so anyone is more than welcome to have a double check).

crawfordsm commented:
Giving it a final review, and everything does look great.  It still needs a changelog entry and could use the mention of the license `# Licensed under a 3-clause BSD style license - see LICENSE.rst`  at the top of the python files.   Happy to merge this once those are added.

The low coverage tests does seem to be with the cython code though the file that it mainly wraps around (bls.c) reports high coverage.

dfm commented:
Hey folks, thanks for the feedback! I've added a changelog entry and the license comments to the headers. I'm also confused about the coverage issues - the file that's bringing down the coverage fraction is auto-generated by Cython. Is there something that I should do to mark that as something that should be skipped? Thanks again!

bsipocz commented:
I'm go ahead and merge this now. Please try to get users testing it, so any fix needed can be done before it goes into the actual release.

bsipocz commented:
Thank you @dfm!

dfm commented:
Thanks @bsipocz and everyone else who gave feedback. It has been great experience contributing to astropy for the first time and I really appreciate how friendly and supportive you have all been!

Also: huge thanks to @mirca who worked with me to develop the initial prototype for this pull request!

bsipocz commented:
@mirca - Retrospect I've noticed that the rebasing and squashing in this PR meant that your commits has been squashed into other ones and thus your author credit has disappeared from the history. I'm truly, terribly sorry about this.
Unfortunately we can't easily fix this in master as we have a strict no-history-rewrite policy there. Could you please @mirca let us know how to proceed, whether you would like us to explore options further?

mirca commented:
@bsipocz that's really kind of you, thanks! But absolutely no worries about that. @dfm already gave me more credit than I definitely deserved :).
